### ISSUE TITLE
Different values for output of prob for every run on Lightened Model C

### ISSUE THREAD

**Issue by sudevbohra** [AlfredXiangWu/face_verification_experiment#75]
I use this to test inference on your model C

`import numpy as np
from collections import OrderedDict
import caffe
def extract_feature(network_proto_path,
                    network_model_path,
                    image_list, layer_name, image_as_grey = True):
    """
    Extracts features for given model and image list.
    Input
    network_proto_path: network definition file, in prototxt format.
    network_model_path: trainded network model file
    image_list: A list contains paths of all images, which will be fed into the
                network and their features would be saved.
    layer_name: The name of layer whose output would be extracted.
    save_path: The file path of extracted features to be saved.
    """
    #network_proto_path, network_model_path = network_path

    net = caffe.Classifier(network_proto_path, network_model_path)

    #caffe.set_mode_gpu()

    # input preprocessing: 'data' is the name of the input blob == net.inputs[0]
    #net.set_mean('data', caffe_root + 'python/caffe/imagenet/ilsvrc_2012_mean.npy')  # ImageNet mean
    #net.set_mean('data', data_mean)
    if not image_as_grey:
        net.set_channel_swap('data', (2,1,0))  # the reference model has channels in BGR order instead of RGB

    #net.set_input_scale('data', 256)  # the reference model operates on images in [0,255] range instead of [0,1]
    #net.set_input_scale('data', 1)

    #img_list = [caffe.io.load_image(p) for p in image_file_list]

    #----- test

    blobs = OrderedDict( [(k, v.data) for k, v in net.blobs.items()])

    #blobs = OrderedDict( [(k, v.data) for k, v in net.blobs.items()])
    shp = blobs[layer_name].shape
    print blobs['data'].shape

    batch_size = blobs['data'].shape[0]
    print blobs[layer_name].shape
    #print 'debug-------\nexit'
    #exit()

    #params = OrderedDict( [(k, (v[0].data,v[1].data)) for k, v in net.params.items()])
    try:
        features_shape = (len(image_list), shp[1],shp[2],shp[3])
    except:
        features_shape = (len(image_list), shp[1])

    features = np.empty(features_shape, dtype='float32', order='C')
    img_batch = []
    for cnt, path in zip(range(features_shape[0]), image_list):
        img = caffe.io.load_image(path, color = not image_as_grey)
        print "IMAGE IS LIKE", img
        if image_as_grey and img.shape[2] != 1:
            img = skimage.color.rgb2gray(img)
            img = img[:, :, np.newaxis]
        if cnt == 0:
            print 'image shape: ', img.shape
        #print img[0:10,0:10,:]
        #exit()
        img_batch.append(img)
        #print 'image shape: ', img.shape
        #print path, type(img), img.mean()
        if (len(img_batch) == batch_size) or cnt==features_shape[0]-1:
            scores = net.predict(img_batch, oversample=False)
            '''
            print 'blobs[%s].shape' % (layer_name,)
            tmp =  blobs[layer_name]
            print tmp.shape, type(tmp)
            tmp2 = tmp.copy()
            print tmp2.shape, type(tmp2)
            print blobs[layer_name].copy().shape
            print cnt, len(img_batch)
            print batch_size
            #exit()
            #print img_batch[0:10]
            #print blobs[layer_name][:,:,0,0]
            #exit()
            '''

            # must call blobs_data(v) again, because it invokes (mutable_)cpu_data() which
            # syncs the memory between GPU and CPU
            blobs = OrderedDict( [(k, v.data) for k, v in net.blobs.items()])

            print '%d images processed' % (cnt+1,)

            #print blobs[layer_name][0,:,:,:]
            # items of blobs are references, must make copy!
            try:
                features[cnt-len(img_batch)+1:cnt+1, :,:,:] = blobs[layer_name][0:len(img_batch),:,:,:].copy()
            except:
                features[cnt-len(img_batch)+1:cnt+1, :] = blobs[layer_name][0:len(img_batch),:].copy()
            img_batch = []

        #features.append(blobs[layer_name][0,:,:,:].copy())
    features = np.asarray(features, dtype='float32')
    print features
    return features

extract_feature("./proto/LightenedCNN_C_deploy.prototxt", "./model/LightenedCNN_C.caffemodel", ["../caffe-tensorflow/face-aligned-ex.jpg"],"prob")`

Every time I run the above script I get different numbers output from the softmax layer. I dug deeper and see that the output from fc2 seems to be changing everytime, all other layers don't change.

Any idea why this happening?

Thank you for your time!

Thank you,
Sudev

**Comment by AlfredXiangWu** [comment]
The deploy.prototxt contains "dropout" layer, which may lead to the phenomenon. You can remove the layer and try again. 

BTW, the model is usually used to extract the "eltwise_fc1" as the feature representation for a face image. It makes no sense to extract the probability of softmax because the identity of the  face image you input may not appear in the training dataset.
